ifErrorDo: should be onErrorDo: and deprecate with transform

astares opened the pull request:
1. Rename #ifErrorDo: into #onErrorDo:
2. use #onErrorDo: in deprecated methods like #ifError: and  #ifErrorDo:
     and move these to Deprecated90 package
3. have a transformation rule for the deprecations in  #ifError: and  #ifErrorDo:
4. Add a test #testOnErrorDo

Fix #6157

astares committed:
ifErrorDo: should be onErrorDo: and deprecate with transform 

1. Rename #ifErrorDo: into #onErrorDo:
2. use #onErrorDo: in deprecated methods like #ifError: and  #ifErrorDo:
     and move these to Deprecated90 package
3. have a transformation rule for the deprecations in  #ifError: and  #ifErrorDo:
4. Add a test #testOnErrorDo

Fix #6157

dionisiydk commented on src/Deprecated90/BlockClosure.extension.st:
```diff
@@ -0,0 +1,15 @@
+{ #category : #'*Deprecated90' }
+BlockClosure >> ifError: errorHandlerBlock [
+
+	self deprecated: 'Use #onErrorDo: instead.' transformWith: '`@receiver ifError: `@arg' -> '`@receiver onErrorDo: `@arg'.
```
the deprecation should more tricky here which is probably the reason why rewrite was not implemented in first PR.
Originally ifError could accept two arguments (using cull). I do not remember exactly. Check Pharo 8.

I think for simplicity we can do rewrite only for zero argument block. Otherwise use deprecation message.

astares committed:
Only rewrite for zero argument blocks

astares committed:
use correct message name in transformation rule

astares commented:
@dionisiydk Can you please check again? Thanks in advance!

dionisiydk commented on src/Deprecated90/BlockClosure.extension.st:
```diff
@@ -0,0 +1,21 @@
+		ifTrue: [ self deprecated: 'Use #onErrorDo: instead.' transformWith: '`@receiver ifError: `@arg' -> '`@receiver onErrorDo: `@arg' ]
+		ifFalse: [ self deprecated: 'Use #onErrorDo: instead.' ].
+		
+	^ self onErrorDo: errorHandlerBlock 
```
You need to wrap it like:
```Smalltalk
self onErrorDo: [:err | errorHandlerBlock cull: err receiver cull: err description]
```
But I did not check in code what exact #cull:cull: it should be to be compatible.

astares replied:
Why wrap? We just dispatch further on. Do you have a problematic case?

dionisiydk replied:
That is original method:
```Smalltalk
ifError: errorHandlerBlock
   ^ self on: Error do: [:ex |
		errorHandlerBlock cull: ex description cull: ex receiver]
```
The issue is about eliminating this strange error unboxing so that the handler block will accept a normal error instance.

dionisiydk replied:
what I am trying to say is that deprecated method should continue work with this strange two parameters

guillep replied:
Yes, the deprecated method should continue working as before :)

guillep commented:
ping?

astares committed:
Merge branch 'Pharo9.0' of https://github.com/pharo-project/pharo into 6157-ifErrorDo-should-be-onErrorDo-and-deprecate-with-transform

astares committed:
Use #on:do: with cull: to fix #6158

astares commented:
@guillep pong!

guillep committed:
Keep deprecation compatibility

Deprecated ifError: and ifErrorDo: should keep compatibility, otherwise they break the idea of a deprecation.
They are called with a block expecting description and receiver as arguments => honor that.
Removed the automatic rewriting as the rewriting rule for that may not be that obvious => let the user decide.

guillep committed:
Do it with ifErrorDo: too

guillep commented:
I've made the deprecated methods work as before to keep compatibility. Let's wait for the CI

guillep reviewed (approved):
The overall changes look ok. Let's wait for the CI.
